nursery/coin_select/src/coin_selector.rs

```diff
@@ -176,440 +174,426 @@ impl<'a> CoinSelector<'a> {
             .sum()
     }
 
-    /// Effective value sum of all selected inputs.
-    pub fn selected_effective_value(&self) -> i64 {
-        self.selected
-            .iter()
-            .map(|&index| self.candidates[index].effective_value(self.opts.target_feerate))
-            .sum()
+    /// The weight of the inputs including the witness header and the varint for the number of
+    /// inputs.
+    fn input_weight(&self) -> u32 {
+        let witness_header_extra_weight = self
+            .selected()
+            .find(|(_, wv)| wv.is_segwit)
+            .map(|_| 2)
+            .unwrap_or(0);
+        let vin_count_varint_extra_weight = {
+            let input_count = self.selected().map(|(_, wv)| wv.input_count).sum::<usize>();
+            (varint_size(input_count) - 1) * 4
+        };
+
+        self.selected_weight() + witness_header_extra_weight + vin_count_varint_extra_weight
```

Comment:
For transaction that have at least one Segwit input, you also need to consider the added witness stack size for all inputs that are spending from a Segwit Script. This is always 1 per legacy input since this stack is always empty.

More information about this here: https://github.com/bitcoin/bitcoin/pull/26567/files#diff-6e06b309cd494ef5da4e78aa0929a980767edd12342137f268b9219167064d13R62-R69.

Reply:
Thanks. I fixed this and added some tests 4e0c907b9d297b7df9e95495dc6ca26d4ecd490e

Interestingly doing this properly broke some prop tests. The prop tester would find edge cases where the branch and bound algorithm found what looked like a good solution with lots of legacy inputs but then it had to add one segwit input and made the weight blow up for the existing inputs. This could probably be fixed but I don't think it's worth the effort and code complexity just to have slightly better selections when you mix legacy and segwit for waste metric etc.